### PR TITLE
Implement Information_Schema with memory table

### PR DESCRIPTION
**Pull request by shenli**
1. Add a MemoryTable in tables. It supports AddRecord, RemoveRecord, Truncate, Row, RowWithCols, Seek. Do not support UpdateRecord and Index.
2. Use MemoryTable to implement Information_Schema. We will drop plans/info.go in the next PR.

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -848,4 +864,6 @@ func FindIndexByColName(t table.Table, name string) *column.IndexedCol {
 
 func init() {
 	table.TableFromMeta = TableFromMeta
+	//driver := localstore.Driver{Driver: goleveldb.MemoryDriver{}}
```
Comment (shenli) [inline table/tables/tables.go]: Remove this.

**Comment by ngaut** [comment]
PTAL @coocood @qiuyesuifeng @nieyy

File: infoschema/infoschema.go [hunk 2]
```diff
@@ -197,9 +199,123 @@ type Handle struct {
...
+		err error
+		tbl table.Table
+	)
+	dbID, err := genGlobalID(store)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	alloc := autoid.NewMemoryAllocator(dbID)
+	nameToTable = make(map[string]table.Table)
+	isTables := make([]*model.TableInfo, 0, 8)
```
Comment (coocood) [inline infoschema/infoschema.go]: `8` -> `len(tableNameToColumns)`

File: infoschema/tables.go [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+	}
+	return &model.ColumnInfo{
+		Name:      model.NewCIStr(col.name),
+		FieldType: fieldType,
+		State:     model.StatePublic,
+	}
+}
+
+func buildTableMeta(tableName string, cs []columnInfo) *model.TableInfo {
+	cols := make([]*model.ColumnInfo, 0, 5)
```
Comment (coocood) [inline infoschema/tables.go]: '5' -> `len(cs)`

File: infoschema/tables.go [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+			rs := fetchColumnsInTable(schema, table)
+			for _, r := range rs {
+				rows = append(rows, r)
+			}
+		}
+	}
+	return rows
+}
+
+func fetchColumnsInTable(schema *model.DBInfo, table *model.TableInfo) [][]interface{} {
```
Comment (coocood) [inline infoschema/tables.go]: rename to `dataForColumnsInTable`.

File: infoschema/tables.go [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+			rs := fetchStatisticsInTable(schema, table)
+			for _, r := range rs {
+				rows = append(rows, r)
+			}
+		}
+	}
+	return rows
+}
+
+func fetchStatisticsInTable(schema *model.DBInfo, table *model.TableInfo) [][]interface{} {
```
Comment (coocood) [inline infoschema/tables.go]: rename to `dataForStatisticsInTable`.

File: meta/autoid/autoid.go [hunk 6]
```diff
@@ -72,6 +72,39 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
 }
 
+var (
+	memID     int64
+	memIDLock sync.Mutex
+)
+
+type memoryAllocator struct {
+	mu    sync.Mutex
+	store kv.Storage
```
Comment (coocood) [inline meta/autoid/autoid.go]: `store` is not used.

File: table/tables/memory_tables.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	return handle, true, nil
+}
+
+// Indices implements table.Table Indices interface.
+func (t *MemoryTable) Indices() []*column.IndexedCol {
+	return nil
+}
+
+// AddIndex implements table.Table AddIndex interface.
+func (t *MemoryTable) AddIndex(idxCol *column.IndexedCol) {
```
Comment (coocood) [inline table/tables/memory_tables.go]: `table.Table` interface has removed some methods including this.

File: infoschema/infoschema.go [hunk 8]
```diff
@@ -197,9 +199,123 @@ type Handle struct {
...
+
+func setColumnID(meta *model.TableInfo, store kv.Storage) error {
+	var err error
+	for _, c := range meta.Columns {
+		c.ID, err = genGlobalID(store)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return err
```
Comment (qiuyesuifeng) [inline infoschema/infoschema.go]: U can directly return nil.

File: infoschema/infoschema.go [hunk 9]
```diff
@@ -241,6 +357,46 @@ func (h *Handle) Set(newInfo []*model.DBInfo, schemaMetaVersion int64) error {
 			}
 		}
 	}
+	// Build Information_Schema
+	info.schemaNameToID[isDB.Name.L] = isDB.ID
+	info.schemas[isDB.ID] = isDB
+	for _, t := range isDB.Tables {
+		tbl, ok := nameToTable[t.Name.L]
+		if !ok {
+			return errors.New("Miss table")
```
Comment (qiuyesuifeng) [inline infoschema/infoschema.go]: Missed table?

File: infoschema/infoschema_test.go [hunk 10]
```diff
@@ -143,10 +144,6 @@ func (*testSuite) TestT(c *C) {
 	c.Assert(ok, IsTrue)
 	c.Assert(col, NotNil)
 
-	col, ok = is.ColumnByID(5)
-	c.Assert(ok, IsFalse)
-	c.Assert(col, IsNil)
-
 	col, ok = is.ColumnByName(dbName, tbName, colName)
```
Comment (qiuyesuifeng) [inline infoschema/infoschema_test.go]: Why remove this test?

File: meta/autoid/autoid.go [hunk 11]
```diff
@@ -72,6 +72,39 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
+	defer alloc.mu.Unlock()
+	if alloc.base == alloc.end { // step
+		memIDLock.Lock()
+		memID = memID + step
+		alloc.end = memID
+		alloc.base = alloc.end - step
+		memIDLock.Unlock()
+	}
+	alloc.base++
+	log.Debugf("[kv] Alloc id %d, table ID:%d, from %p, database ID:%d", alloc.base, tableID, alloc, alloc.dbID)
```
Comment (qiuyesuifeng) [inline meta/autoid/autoid.go]: Should add memoryAllocator information to debug log?

File: table/tables/memory_tables.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+// RowWithCols implements table.Table RowWithCols interface.
+func (t *MemoryTable) RowWithCols(ctx context.Context, h int64, cols []*column.Col) ([]interface{}, error) {
+	if h >= int64(len(t.rows)) || h < 0 {
+		return nil, errors.New("Can not find the row")
+	}
+	row := t.rows[h]
+	if row == nil {
+		return nil, errors.New("Can not find row")
```
Comment (qiuyesuifeng) [inline table/tables/memory_tables.go]: Should unify this error and error in line 171.

File: table/tables/memory_tables_test.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (qiuyesuifeng) [inline table/tables/memory_tables_test.go]: Copyright 2016

File: table/tables/memory_tables_test.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+	alloc := autoid.NewMemoryAllocator(int64(10))
+	ts.tbl, _ = tables.MemoryTableFromMeta(alloc, tblInfo)
+}
+
+func (ts *testMemoryTableSuite) TestMemoryBasic(c *C) {
+	ctx := ts.se.(context.Context)
+	tb := ts.tbl
+	c.Assert(tb.Meta().ID, Greater, int64(0))
+	c.Assert(tb.Meta().Name.L, Equals, "t")
+	c.Assert(tb.Meta(), NotNil)
```
Comment (qiuyesuifeng) [inline table/tables/memory_tables_test.go]: Seems useless.

File: meta/autoid/autoid.go [hunk 15]
```diff
@@ -72,6 +72,39 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
+	if tableID == 0 {
+		return 0, errors.New("Invalid tableID")
+	}
+	alloc.mu.Lock()
+	defer alloc.mu.Unlock()
+	if alloc.base == alloc.end { // step
+		memIDLock.Lock()
+		memID = memID + step
+		alloc.end = memID
+		alloc.base = alloc.end - step
```
Comment (zimulala) [inline meta/autoid/autoid.go]: The above three lines of code can be understood as the following code ?
`alloc.end = memID + step
 alloc.base = memID`

File: infoschema/tables.go [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+
+const (
+	tableSchemata      = "SCHEMATA"
+	tableTables        = "TABLES"
+	tableColumns       = "COLUMNS"
+	tableStatistics    = "STATISTICS"
+	tableCharacterSets = "CHARACTER_SETS"
+	tableCollations    = "COLLATIONS"
+	tableFiles         = "FILES"
+	catalogVal         = "def"
```
Comment (zimulala) [inline infoschema/tables.go]: def -> DEF ?

File: infoschema/infoschema_test.go (hunk 10 above)
Comment (coocood) [inline infoschema/infoschema_test.go]: The column ID 5 is allocated by `infoschema` tables, so the test will not pass.

File: infoschema/tables.go (hunk 16 above)
Comment (coocood) [inline infoschema/tables.go]: It's is lower case 'def' in MySQL

File: meta/autoid/autoid.go (hunk 15 above)
Comment (coocood) [inline meta/autoid/autoid.go]: memID is a global variable, it needs to be updated.

Comment (zimulala) [inline meta/autoid/autoid.go]: yep, but store and allocator is one-to-one， so maybe `memID` is needless？

File: infoschema/tables.go (hunk 16 above)
Comment (zimulala) [inline infoschema/tables.go]: Gotcha

File: meta/autoid/autoid.go (hunk 15 above)
Comment (coocood) [inline meta/autoid/autoid.go]: Allocator is used by a single table to allocate recordID.

Comment (zimulala) [inline meta/autoid/autoid.go]: I means `memoryAllocator` only initializes in `initMemoryTables`， so maybe we can replace `memID` with `alloc.base`？

**Comment by nieyy** [comment]
I have two questions:
1. All perfschema tables support `SELECT` operations, but some tables support different operations, such as, `setup_timers` table support `UPDATE` only, `events_xxx` tables support `TRUNCATE` only, how to code w/ memory tables?
2. How to use `table.Table` interface to simulate `UPDATE` operation?

**Comment by coocood** [comment]
@nieyy 
For infrequent updates, you can execute SQL statement.
For frequent update, you can maintain the handles of the added records, use that to call `table.Update` method.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@nieyy we can later implement another table implementation which optimized for instruments.

**Comment by nieyy** [comment]
LGTM，after discussed w/ @shenli , I will add another special memory table for events_xxx tables once the instrumentation efficiency is really a problem.

File: table/tables/memory_tables.go [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+	"github.com/ngaut/log"
+	"github.com/petar/GoLLRB/llrb"
+	"github.com/pingcap/tidb/column"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/types"
+	"sync"
```
Comment (shenli) [inline table/tables/memory_tables.go]: Move before "github.com/xxx"

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**